docs/rfcs/0007-flow-modules.md

```diff
@@ -0,0 +1,295 @@
+* [Configuration blocks](https://grafana.com/docs/agent/latest/flow/reference/config-blocks/#configuration-blocks) will not be supported. 
+* Arguments and exports within a module must be unique
+
+## Proposal
+
+Add the ability to load `modules` as subgraphs to the primary `graph`. Modules may call other modules within a reasonable stack size depth. Modules are represented as a river string that is interpreted with a defined set of arguments and exports.
+
+The initial component will be `module.single` that will load a single module. Internally these modules will be namespaced so they cannot affect children or parent graphs except via arguments and exports.
+
+Modules will have access to any standard function and any other component exempting singletons. Internally each component in the module will have an `id` that is prepended with the parent's `id` for identification purposes outside of the module. Within the module a component can reference another sibling component normally. There are no known limits on the datatype that a module can use as an argument or export.
+
+`modules.multiple` will use the key of the `map(string)` to uniquely identify a specific module.
+
+### Component Options
+
+
+Given the above example, the `id` of `integrations.mysql "server1"` would be `module.single.mysql.integrations.mysql.server1`. The `data-agent` field would also be prefixed. There are some inherent issues, deeply nested metrics are likely to run into Prometheus label value limits. On Windows platforms there could be issues with the `data-agent` length. These are issues that currently exist in Agent Flow but are more easily hit using deeply nested modules.
+
+
+### Failure Modes
+
+#### Option 1: When a module fails then fail itself and children.
+
+If an error occurs while re-evaluating a module then the module marks itself as unhealthy and unloads the original module and all children.
+
+*Pros*
+
+* Simple to implement
+* Easy to understand
+
+*Cons*
+
+* One failure mode can cascade
+
+#### Option 2: Modules keep last good value
+
+If an error occurs while re-evaluating a module then the module marks itself as unhealthy and attempts to keep the original module. This may have an issue with cascading failures, if a module depends on a module then the system may enter an inconsistent state while applying and then rolling back the change.
+
+For example, `Module A` has two sub-modules `Module B` and `Module C`. During reevaluation `Module B` reloads appropriately but `Module C` fails. `Module A` unloads both modules and then reloads the last good string. In the case that the last good string also fails then `Module A` is unhealthy and non-functional and `Module A's` submodules do not exist.
+
+*Pros*
+
+* Allows more resilient usage
+
+*Cons*
+
+* Can create undefined behavior
+* Complex to unload and reload
+
+## `modules.multiple`: Allowing multiple modules to be loaded at once
```

Comment:
It feels like there can be a ton of discussion around this component alone. What do you think about leaving it out of the scope of the modules RFC and just starting with one way to load modules?

Reply:
I agree with this, mainly since the multi-use case introduces a lot of additional complexity that doesn't necessarily need to be discussed up front with the idea of modules. Flow describes itself as[ being made up of components](https://grafana.com/docs/agent/latest/flow/concepts/components/)that handle a single task (which of course _can_ go very deep, but thus far most components scope is not huge by any means).

Modules take this one step further and provide a wrapper over multiple components to give you a black box of sorts that does something more interesting/common use case/etc, which is great! Multi-module support feels like sugar on top of that for when you want to bundle _even more_, instead of just listing all the individual pieces together and keeping track of a large list. That's also great, but feels like jumping the gun while we're still figuring out what it means to load a _single module_; sugar before the tea, cart before the horse, and all that :^) 

That's just my 2c, but I 100% agree the multi-module aspect should be moved into its own RFC once what a module is has been more solidified.

Reply:
Agreed with the above and removing `modules.multiple`.